pyquil/api/_compiler.py

```diff
@@ -353,3 +380,71 @@ def reset(self) -> None:
         timeout = self.client.timeout
         self.client.close()
         self.client = Client(self.endpoint, timeout=timeout)
+
+
+@dataclass
+class HTTPCompilerClient:
+    """
+    A class which partially implements the interface of rpcq.Client, to allow the QPUCompiler to
+    send compilation requests over HTTP(S) rather than ZeroMQ.
+
+    :param endpoint: the base url to which rpcq methods will be appended.
+    :param session: the ForestSession object which manages headers and authentication.
+    """
+
+    endpoint: str
+    session: ForestSession
+
+    def call(self,
+             method: str,
+             payload: Optional[Message] = None,
+             *,
+             rpc_timeout: int = 30) -> Message:
+        """
+        A partially-compatible implementation of rpcq.Client#call, which allows calling rpcq
+        methods over HTTP following the scheme:
+
+            POST <endpoint>/<method>
+            body: json-serialized <payload>
+
+        This implementation is meant for use only with the QPUCompiler and is not intended to be
+        a fully-compatible port of rpcq from ZeroMQ to HTTP.
+
+        If the request succeeds (per HTTP response code), the body of the response is parsed into 
+        an RPCQ Message.
+
+        If the request fails, the response body should be a JSON object with a ``message`` field
+        indicating the cause of the failure. If present, that message is delivered to the user.
+
+        :param payload: the rpcq message body
+        :param rpc_timeout: the number of seconds to wait to read data back from the service
+                            after connection.
+                            @see https://requests.readthedocs.io/en/master/user/advanced/#timeouts
+        """
+        url = urljoin(self.endpoint, method)
+
+        if payload:
+            body = to_json(payload)
+        else:
+            body = None
+
+        response = self.session.post(url, json=body, timeout=(1, rpc_timeout))
+
+        try:
+            response.raise_for_status()
+        except RequestException as e:
+            message = f'QPU Compiler {method} failed: '
+
+            try:
+                contents = response.json()
+            except Exception:
```

Comment:
should we be more specific or is it too complicated?

Reply:
I didn't want any failure in the parsing of the error response to mask the fact that there was an error in a call to the compiler itself